processing/src/main/java/io/druid/segment/IndexMaker.java

```diff
@@ -1011,47 +987,13 @@ public VSizeIndexedInts apply(List<Integer> input)
 
           dictionary = GenericIndexed.fromIterable(
               Iterables.concat(nullList, dimensionValues),
-              GenericIndexed.stringStrategy
+              GenericIndexed.STRING_STRATEGY
           );
           multiValCol = null;
-          singleValCol = new AbstractList<Integer>()
-          {
-            @Override
-            public Integer get(int index)
-            {
-              Integer val = vals.get(index);
-              if (val == null) {
-                return 0;
-              }
-              return val + 1;
-            }
-
-            @Override
-            public int size()
-            {
-              return vals.size();
-            }
-          };
+          singleValCol = new NullsAtZeroConvertingIntList(vals);
         } else {
           multiValCol = null;
-          singleValCol = new AbstractList<Integer>()
-          {
-            @Override
-            public Integer get(int index)
-            {
-              Integer val = vals.get(index);
-              if (val == null) {
-                return 0;
-              }
-              return val;
-            }
-
-            @Override
-            public int size()
-            {
-              return vals.size();
-            }
-          };
+          singleValCol = new NullsAtZeroConvertingIntList(vals);
```

Comment:
It turns out that the "dim value is null but dimension doesn't already have a null entry for a single value column" is the required case to reach this code. And I'm having a hard time getting this case to occur "naturally".